### PR TITLE
Inplace version of wikize refs

### PR DESCRIPTION
**Pull request by markcmiller86**
Addresses/related to #702, #695, #363

### CODE REVIEW

**Comment by markcmiller86** [comment]
Two diagrams I am adding in this PR conversation describe how the new, in-place version of wikize-refs should work...

Slide1

Slide2

### Some Comments

* The document structured can be viewed in four blocks...
   * The main content, including any bssw.io metadata block(s), *above* link definitions.
   * The author's link definitions block (which will be *hidden* from markdown processing by XML comment bracketing)
   * The placeholder link definitions block (lines of the form `^[J]: #refJ`) for internal links to the table of references
   * The table of references (lines of the form `^<a name="refJ"></a>J | [`) for actual links to referenced content
* The link definitions as *authored* all appear together in a block at the *bottom* of the file, *after* the main content and even after any bssw.io metadata blocks. Thus, in line-by-line processing, upon encountering the *first* such link definition, we can assume all the document's main content has already been seen.
* The bracketing xml comments `<!--- BEGIN LINK DEFS` and `END LINK DEFS --->` are needed to remove the author's set of link definitions from markdown processing (e.g. to *hide* them from any markdown processor).
* The wikize refs operation will renumber the author's link definitions `1...N` and will update any footnote refs (e.g. `<sup>[K]</sup>` in the main content accordingly.
   * **NOTE**: Any link defintion of the form `^[J]: https?://` (`^` having regex meaning for beginning of line and `?` having regex meaning for previous character appearing zero or one times and where `J` is an integer)  encountered *outside* of the author's link definitions block will get treated like it is a new (since last processing the file) link definition to be included in the newly updated author's link definitions block.
* Placeholder link definition lines of the form `^[J]: #refJ` will be removed. The wikize operation will output these anew each invocation.
* Reference table lines of the form `^<a name="refJ"></a>J | [` will be removed. The wikize operation will output these anew each invocation.

**Comment by markcmiller86** [comment]
Ok, I've got a working in-place version of this tool. I've added features to it too.

* Repeated reruns will not change file if the file hasn't been edited
* Modularized it a lot more
* Documented the code quite a bit more
* Handled some special cases like a doc using GFM link definition lists for *body* content in addition to *footnoted* references.
* Found a problem with my use of re module and fixed it.
* Adjusted usage and CL args
* 

I've tested it many different ways manually. I should *write* a test for it but haven't.

I think this is ready to go.

**Comment by bartlettroscoe** [comment]
@markcmiller86, I pulled and checked out this branch and tried to run it on my already-wikized blog article:

* https://github.com/betterscientificsoftware/bssw.io/blob/01065b43dc8556c72694bf5426a2938828faef55/Articles/Blog/WhenNotToWriteAutomatedTests.md

and it produced a new file when I asked it to run 'in-place':

```
[rabartl@s1049009 Blog (mcm86-19feb21-inplace-wikize-refs)]$ git status
On branch mcm86-19feb21-inplace-wikize-refs
Your branch is up to date with 'github/mcm86-19feb21-inplace-wikize-refs'.

nothing to commit, working tree clean

[rabartl@s1049009 Blog (mcm86-19feb21-inplace-wikize-refs)]$ ./wikize_refs.py -i WhenNotToWriteAutomatedTests.md 

[rabartl@s1049009 Blog (mcm86-19feb21-inplace-wikize-refs)]$ git status
On branch mcm86-19feb21-inplace-wikize-refs
Your branch is up to date with 'github/mcm86-19feb21-inplace-wikize-refs'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        WhenNotToWriteAutomatedTests-wikized.md

nothing added to commit but untracked files present (use "git add" to track)
``` 

Why did it create the new file `WhenNotToWriteAutomatedTests-wikized.md` when it was asked to process the file in-place?  Is that expected.  (I can dig into the code my self but it is worth asking if that is expected behavior.)

We should write some automated tests to pin down what this tool is supposed to do.  I am willing to write the beginnings of these tests using a simple Python unit test driver based on the standard unittest Python module. 

Might I suggest:

* Move the `wikize_refs.py` script into a new subdir `bssw.io/utils/` (we can put a symlink to this under `bssw.io/Artilces/Blob/` to make easy to find still).
* Create a starting `wkize_refs_tests.py` test driver in `bssw.io/utils/`.
* Change from ` optparse` to `argparse` (which is supported in Python 2.7 and 3.x and creates a nicer help message).
* Update documentation to match desired/expected behavior.

I can do all of this in a branch based off this branch and then post a PR against this PR if you are okay with that?

**Comment by markcmiller86** [comment]
> and it produced a new file when I asked it to run 'in-place'

Whoops. There is a logic error with that and option to support `-o`. I see it and will fix. Thx

**Comment by markcmiller86** [comment]
@bartlettroscoe I just pushed an update that fixes the issue with in place behavior.

**Comment by markcmiller86** [comment]
> Might I suggest:
> 
> * Move the `wikize_refs.py` script into a new subdir `bssw.io/utils/` (we can put a symlink to this under `bssw.io/Artilces/Blob/` to make easy to find still).
> * Create a starting `wkize_refs_tests.py` test driver in `bssw.io/utils/`.
> * Change from ` optparse` to `argparse` (which is supported in Python 2.7 and 3.x and creates a nicer help message).
> * Update documentation to match desired/expected behavior.

This is all fine too...maybe also change name of tool to `wikize-refs.py` (using hyphen) instead of `wikize_refs.py` (using underscore). FWIW, I punted on `optparse`/`argparse` because it was already there and working.

**Comment by bartlettroscoe** [comment]
> maybe also change name of tool to `wikize-refs.py` (using hyphen) instead of `wikize_refs.py`

@markcmiller86, if you do that, then you can't unit test the code in `wikize-refs.py` because that is not a valid module name.  So we need to leave it as `wikize_refs.py` with an underscore for unit testing purposes.  But we can always add a symlink from `wikize-refs.py` to `wikize_refs.py` if people prefer that name as well.  And for that matter just call the symlink `wikize-refs` (since users should not care if it is implement in Python or Perl or Bash, etc.).

**Comment by bartlettroscoe** [comment]
> This is all fine too

Okay, I will put in a PR for the suggested changes and the beginning of some automated tests.

> FWIW, I punted on optparse/argparse because it was already there and working.

It is not a big deal.  But it is a pretty easy translation.  I will give it a shot in my PR.

**Comment by markcmiller86** [comment]
FYI...I've made it work as a module a little better...

```
>>> import wikize_refs
>>> opts={'warn':True, 'in_place':True, 'outfile':'foobar.md'}
>>> opts
{'warn': True, 'in_place': True, 'outfile': 'foobar.md'}
>>> wikize_refs.main(opts, "PullRequestSizeConsiderations-wikized.md")
>>> ^D
% ls foobar.md 
foobar.md
% diff -s PullRequestSizeConsiderations-wikized.md foobar.md
Files PullRequestSizeConsiderations-wikized.md and foobar.md are identical
```

**Comment by markcmiller86** [comment]
This makes me think...I should add a feature such that if output is unchanged from input, we should report that *and* not do anything that would change input file (e.g. even timestamp, modified time, etc.).

**Comment by markcmiller86** [comment]
@bartlettroscoe ... BTW, you will notice two files, `ReferencesInMarkdownHybridApproach.md` and `ReferencesInMarkdownHybridApproach-wikized.md` here. That is intentional in order to show before/after comparison of running `wikize_refs.py` for the example text that is used there.

File: Articles/Blog/test_wikizer.md [hunk 1]
```diff
@@ -40,13 +40,25 @@ Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia
...
+
+### Ref with title but no bibliographic
+
+### Ref with no title but bibliographic (is this nonsensical)
+
+### Ref with http:// vs https://
+
+### Refs in XML comment block should be ignored (except the one wikize_refs recognizes)
+
+### Can we detect main content *after* link def block and flag it
```
Comment (bartlettroscoe) [inline Articles/Blog/test_wikizer.md:61]: @markcmiller86, what is this file for?  Is this a set of todos for setting up unit tests?

Comment (markcmiller86) [reply Articles/Blog/test_wikizer.md:61]: Yes

**Comment by markcmiller86** [comment]
Thinking in terms of making this tool more broadly useful to the community as a whole, a few additional features occured to me that I will just capture here in this issue so I don't forget about them.

* support reference *databases*...basically independent `.md` files with a slew of link definition lines in them.
   * Supporting this would mean footnote refs in the main content would have to search first the originating document and then the dbs for a match and then output matching link definition lines.
* support for journal-style refs (e.g. `([miller97],[bartlett02])`)
* a short-hand for `<sup>...</sup>` that the tool itself recognizes
* maybe handle different variations of markdown

**Comment by bernhold** [comment]
@markcmiller86 unless you have other uses for this, I encourage you not to get too far ahead of what we actually need and will use in BSSw.  I think the References Working Group needs to complete its work to figure out how we want to proceed.  But offhand, I would say that databases seem unlikely to be useful in BSSw (since I think we don't want to encourage references in general), and journal-style references would depend on the WG recommendations.  Likewise variations of Markdown don't really apply to BSSw either.

**Comment by bartlettroscoe** [comment]
@markcmiller86, the journal-style references and a reference DB would be pretty useful longer term but likely not needed for BSSw.  But the journal-style references would be pretty useful since the names are more meaningful than `ref<N>`.

**Comment by bartlettroscoe** [comment]
> I would say that databases seem unlikely to be useful in BSSw (since I think we don't want to encourage references in general)

@bernhold, I guess we need to come to a consensus on references.  I gave my views on this in https://github.com/betterscientificsoftware/bssw.io/issues/702#issuecomment-778466528.  (But with in-place edits, almost all of the disadvantages of generated files goes away.)

**Comment by markcmiller86** [comment]
Sheesh...I thought this disclaimer...

>I will just capture here in this issue so I don't forget about them.

was sufficient to indicate my intention...to capture longer term ideas. I guess not. Sorry if I scared you.

I think this tool and the features it supports could be useful to the wider community, not just bssw.io. I am thinking in terms of moving this to its own repo that I host and maintain apart from anything I am doing here (as was mentioned by @bartlettroscoe in another issue here)

**Comment by bartlettroscoe** [comment]
@markcmiller86, let me know when you are ready for me to take a deeper look and start setting up automated tests.  Getting some system-level tests in place first should be the first priority.

**Comment by bernhold** [comment]
> @bernhold, I guess we need to come to a consensus on references. I gave my views on this in [#702 (comment)](https://github.com/betterscientificsoftware/bssw.io/issues/702#issuecomment-778466528). (But with in-place edits, almost all of the disadvantages of generated files goes away.)

Ross, indeed, the hope was that the working groups would meet/discuss and converge on recommendations to the full EB so that we can lay these things to rest.

**Comment by bartlettroscoe** [comment]
FYI: I pulled this updated branch and used it to update my blog article "When NOT to write tests".  I posted this in the new branch:

*  https://github.com/bartlettroscoe/bssw.io/blob/360-718-wikize-when-not-to-write-tests/Articles/Blog/

with the changes shown [here](https://github.com/betterscientificsoftware/bssw.io/compare/mcm86-19feb21-inplace-wikize-refs...bartlettroscoe:360-718-wikize-when-not-to-write-tests?expand=1) and the updated document at:

* https://github.com/bartlettroscoe/bssw.io/blob/360-718-wikize-when-not-to-write-tests/Articles/Blog/WhenNotToWriteAutomatedTests.md

I tested adding a reference and regenerating and it worked great.

Next I will post a PR that does the things mentioned [above](https://github.com/betterscientificsoftware/bssw.io/pull/718#issuecomment-786076319) that have not already been done.

**Comment by bartlettroscoe** [comment]
> Next I will post a PR that does the things mentioned [above](https://github.com/betterscientificsoftware/bssw.io/pull/718#issuecomment-786076319) that have not already been done.

@markcmiller86, I posted the new PR #727 against this PR.  Can you take a look?  Also, there was a use case related to Handle an isoloated -o / --outfile (which means to use default name of output file)" that I could not figure our and is currently disabled.

**Comment by bartlettroscoe** [comment]
@markcmiller86, with the merge of PR #727 to this PR's branch, are you okay to merge this PR?  Since the `wikize_refs.py` tool is not in heavy usage the risk seems pretty low.  Then we can make more changes, add more tests, etc. in new PRs (as per [PR size matters](#670)).  (In fact, I will post a PR that adds the beginnings of `wikize_refs_unit_tests.py` as the next PR after this PR merges.)

**Comment by bartlettroscoe** [comment]
FYI: Not a big deal for a fast-running utility like this but with ctest, you can run the tests in parallel and get some good speedup.  On my Windows 10 laptop under cygwin, here is running the tests one at at time:

```
[rabartl@s1049009 build_tests (mcm86-19feb21-inplace-wikize-refs)]$ ctest -j1
Test project /home/rabartl/Publications/Publications/bssw.io/utils/build_tests
    Start 1: wikize_refs_help
1/7 Test #1: wikize_refs_help .............................   Passed    0.29 sec
    Start 2: wikize_refs_no_args_pass
2/7 Test #2: wikize_refs_no_args_pass .....................   Passed    0.30 sec
    Start 3: wikize_refs_no_args_fail
3/7 Test #3: wikize_refs_no_args_fail .....................   Passed    0.26 sec
    Start 4: wikize_refs_in_place
4/7 Test #4: wikize_refs_in_place .........................   Passed    0.39 sec
    Start 5: wikize_refs_in_place_no_save_orig
5/7 Test #5: wikize_refs_in_place_no_save_orig ............   Passed    0.32 sec
    Start 6: wikize_refs_explicit_out
6/7 Test #6: wikize_refs_explicit_out .....................   Passed    0.37 sec
    Start 7: wikize_refs_in_place_and_explicit_out_fail
7/7 Test #7: wikize_refs_in_place_and_explicit_out_fail ...   Passed    0.31 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) =   2.28 sec
```

and here is running them in parallel:

```
[rabartl@s1049009 build_tests (mcm86-19feb21-inplace-wikize-refs)]$ ctest -j8
Test project /home/rabartl/Publications/Publications/bssw.io/utils/build_tests
    Start 7: wikize_refs_in_place_and_explicit_out_fail
    Start 5: wikize_refs_in_place_no_save_orig
    Start 4: wikize_refs_in_place
    Start 6: wikize_refs_explicit_out
    Start 2: wikize_refs_no_args_pass
    Start 1: wikize_refs_help
    Start 3: wikize_refs_no_args_fail
1/7 Test #7: wikize_refs_in_place_and_explicit_out_fail ...   Passed    0.42 sec
2/7 Test #5: wikize_refs_in_place_no_save_orig ............   Passed    0.49 sec
3/7 Test #6: wikize_refs_explicit_out .....................   Passed    0.47 sec
4/7 Test #4: wikize_refs_in_place .........................   Passed    0.58 sec
5/7 Test #3: wikize_refs_no_args_fail .....................   Passed    0.35 sec
6/7 Test #1: wikize_refs_help .............................   Passed    0.40 sec
7/7 Test #2: wikize_refs_no_args_pass .....................   Passed    0.47 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) =   0.74 sec
```

A 3x speedup is not terrible for a fast running io utility like this.

**Comment by markcmiller86** [comment]
Yes ok to merge

Sent from my iPhone

On Mar 3, 2021, at 9:17 AM, Roscoe A. Bartlett <notifications@github.com> wrote:

﻿

@markcmiller86<https://urldefense.us/v3/__https://github.com/markcmiller86__;!!G2kpM7uM-TzIFchu!i731X7A7jTGhSh98RH94QVWrhfFauTcyLYk95cd3S4B0j4f3LOQJNpS1E-McINEHkw$>, with the merge of PR #727<https://urldefense.us/v3/__https://github.com/betterscientificsoftware/bssw.io/pull/727__;!!G2kpM7uM-TzIFchu!i731X7A7jTGhSh98RH94QVWrhfFauTcyLYk95cd3S4B0j4f3LOQJNpS1E-Mk83hv-w$> to this PR's branch, are you okay to merge this PR? Since the wikize_refs.py tool is not in heavy usage the risk seems pretty low. Then we can make more changes, add more tests, etc. in new PRs (as per PR size matters). (In fact, I will post a PR that adds the beginnings of wikize_refs_unit_tests.py as the next PR after this PR merges.)

**Comment by bartlettroscoe** [comment]
Addresses #363, #702

**Comment by JeanShuler** [comment]
Thanks for taking care of this!

**Comment by bartlettroscoe** [comment]
Related to my epic SEPW-211

**Pull request merged**